aten/src/ATen/core/IList.h

```diff
@@ -0,0 +1,258 @@
+#pragma once
+
+#include <ATen/core/List.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Exception.h>
+
+#include <initializer_list>
+#include <iterator>
+#include <type_traits>
+
+namespace c10 {
+// Applies arbitrary macros to each `IListTag`.
+#define TORCH_ILIST_FORALL_TAGS(_, ...) \
+  _(Unboxed, ##__VA_ARGS__)             \
+  _(Boxed, ##__VA_ARGS__)
+
+// Builds the name of the implementation class for `TAG`.
+#define TORCH_ILIST_IMPL(T, TAG) \
+  c10::detail::IListTagImpl<T, c10::IListTag::TAG>
+
+// Defines a "switch-case" for `TAG`. Inside, it executes `BODY`,
+// while bringing to scope:
+//     - `ImplT`: the implementation class for `TAG`
+//     - `this_`: the result of unwrapping `this`
+#define TORCH_ILIST_UNWRAP_CASE(TAG, BODY)  \
+  case c10::IListTag::TAG: {                \
+    using ImplT = TORCH_ILIST_IMPL(T, TAG); \
+    auto& this_ = ImplT::unwrap(*this);     \
+    BODY                                    \
+  } break;
+
+// Dispatches the unwrap call, depending on `TAG`, followed by
+// the execution of `BODY`. It aborts if `TAG` is not a `IListTag`.
+#define TORCH_ILIST_UNWRAP(TAG, BODY)                      \
+  switch (TAG) {                                           \
+    TORCH_ILIST_FORALL_TAGS(TORCH_ILIST_UNWRAP_CASE, BODY) \
+    default:                                               \
+      TORCH_INTERNAL_ASSERT(false, "invalid IList tag.");  \
+  }
+
+enum class IListTag {
+#define DEFINE_TAG(tag, ...) tag,
+  TORCH_ILIST_FORALL_TAGS(DEFINE_TAG)
+#undef DEFINE_TAG
+      None
+};
+
+namespace detail {
+template <typename T>
+using IListConstRef = typename ivalue_to_const_ref_overload_return<T>::type;
+
+/*
+ * Interface that implements key functions for each `IListTag` type.
+ *
+ * You should create an specialization of this class for each
+ * possible combination of `IListTag` type (except `None`) and
```

Comment:
Maybe just enumerate the current full list here, for clarity:
`(at::Tensor, IListTag::unboxed)`
`(at::Tensor, IListTag::boxed)`
`(at::OptionalTensorRef, IListTag::unboxed)`
`(at::OptionalTensorRef, IListTag::boxed)`